### PR TITLE
Integrate ComputedTransforms into the DataSpec System

### LINKED ISSUE #4205: Add the ComputedTransforms to the DataSpec System
Currently, to enable a Transform to be performed, the user must actively click on a control to force the calculation.  Integrating this into the DataSpec system will allow for the transformation to be performed on the fly without the need for a user to force the calculation.

### PR DESCRIPTION
**Pull request by justacec**
This PR will integrate the new Computed Transforms into the DataSpec system.  This will allow the user to apply transforms to the data provided to various plotting functions.

As an example usage, the following line will apply a random jitter to the x coordinate of a set of circles

    bk.circle(x={'field': 'xvalues', 'transform': Jitter(width=0.3)}, y='yvalues', source=mydata)

- [ ] issues: fixes #4205 
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

This is no where near ready but I wanted to put the PR in to begin the necessary work.

### CODE REVIEW

File: bokeh/core/properties.py [hunk 1]
```diff
@@ -1600,7 +1600,7 @@ def set_from_json(self, obj, json, models=None):
 
 class DataSpec(Either):
     def __init__(self, typ, default, help=None):
-        super(DataSpec, self).__init__(String, Dict(String, Either(String, typ)), typ, default=default, help=help)
+        super(DataSpec, self).__init__(String, Dict(String, Either(String, Instance(HasProps), typ)), typ, default=default, help=help)
```
Comment (bryevdv) [inline bokeh/core/properties.py]: I think this can be more specific, e.g. `Instance(Transform)`

**Comment by bryevdv** [comment]
holy crap this just works:
```
from bokeh.models.transforms import Jitter
from bokeh.plotting import 

output_file("/tmp/foo.html")

source = ColumnDataSource(data=dict(x=[1]*100, y=list(range(100))))

p = figure()
p.circle(x={'field': 'x', 'transform': Jitter(width=0.3)}, y='y', source=source)
p.circle(x=[0,2], y=[50, 50])

show(p)
```

screen shot 2016-05-10 at 2 19 10 pm

ping @pzwang !

**Comment by bryevdv** [comment]
So then a next improvement could be simple convenience functions:
```
def jitter(field, width):
    return {'field': field, 'transform': Jitter(width=width)}

p = figure()
p.circle(x=jitter('x', 0.3), y='y', source=source)
p.circle(x=[0,2], y=[50, 50])

show(p)

```

This is exciting @justacec !

**Comment by bryevdv** [comment]
OK this is awesome! 
```
output_file("/tmp/foo.html")
source = ColumnDataSource(data=dict(x=np.linspace(0, 10, 100)))
interp = StepInterpolator(x=list(range(10)), y=list(range(10)))
p = figure()
p.circle(x='x', y={'field': 'x', 'transform': interp}, source=source)
show(p)
```
screen shot 2016-05-10 at 2 40 14 pm

but also maybe a small issue to check out? Wondering about that last set of points on the right down at `y=0

**Comment by bryevdv** [comment]
This case will require a little attention too:
```
source = ColumnDataSource(data=dict(y=list(range(1000))))
p = figure()
p.circle(x={'value': 1, 'transform': Jitter(width=0.3)}, y='y', source=source)
p.circle(x=[0,2], y=[50, 50])
show(p)
```
screen shot 2016-05-10 at 2 57 52 pm

It "works" but only jitters the one single value and uses that everywhere. I think the desirable semantics will be to "promote" values transformed this way to full arrays (or perhaps a the transform can specify if that should happen or not)

**Comment by justacec** [comment]
Glad you enjoy it.  I got side tracked because of something.  I need to get back over to it.  I will see if I can carve out some time this upcoming weekend to look at things.

**Comment by bryevdv** [comment]
@justacec FYI another option would be to re-open this PR as a branch on the main repo instead of your fork. Then I could help out a bit on moving it forward.

**Comment by justacec** [comment]
I thought I did....

**Comment by bryevdv** [comment]
It looks like its coming from your personal fork?

screen shot 2016-05-23 at 9 44 24 am

**Comment by justacec** [comment]
Humm.  I will check into that

**Comment by bryevdv** [comment]
@justacec If it's OK, I am going to open a new PR on the main repo to supersede this one?

**Comment by justacec** [comment]
Bryan,

Sure.  I am really super confused about the branch though.  Instead of
creating the branch from my fork I created it from the bokeh repo by
clicking on the branch pull down from the main bokeh github page and
created a new branch.  I then pushed some changes and then created a pull
request from that to support discussions.  I thought that would allow you
to make changes to the branch as well.  Are you sure you can not switch
that that branch and make changes as is?

That being said.  I have been meaning to get ahold of you to let you know
to go ahead and make changes.  Things have just been really busy for me and
I find it very hard to find time for weekend coding.  I look forward to
seeing what comes of this.

Justace

**Comment by justacec** [comment]
As an aside, if you figure out what I did wrong when I created the branch I
would like if you could let me know.

**Comment by bryevdv** [comment]
@justacec I see the branch on the main repo, but it looks like it doesn't actually have any commits in it:

screen shot 2016-06-01 at 3 17 07 pm

Which is why no pull request was made for it (even if you tried to make one, GH will not let you make a PR from an empty branch) If you'd like try and push new changes to that branch, and re-try making a PR for it, I can wait. I'd definitely like for your contributions to be acknowledged.

**Comment by justacec** [comment]
roger.  Let me look into this tonight.  I have to finish a block of code
right now and then hit the gym.

**Comment by bryevdv** [comment]
great!

**Comment by justacec** [comment]
@bryevdv Ok, I think that I might have fixed it up.  I had to push my repo to upstream and then do the PR.  The updated PR is #4420.  Lets migrate to there.  (I forgot how much fun this project was...)

**Pull request closed without merging**